src/parser.ts

```diff
@@ -183,5 +257,13 @@ export { version };
 const version = packageJSON.version;
 
 export function parse(code: string, options: ParserOptions) {
```

Comment:
I think it would be good to keep the parse() function as being as simple as possible:

parse(CODE, OPTIONS): AST

Maybe we could add an alternative method:

parseAndGenerateServices(CODE, OPTIONS): YOUR INTERFACE

Then the opt-in is even more explicit for consumers, and the parse function interface will match what other parsers in this space expose.

WDYT?

Reply:
You could also put the parser services in the AST.

Reply:
That would make them harder to serialize though right?

Reply:
I'm fine with adding a second method for explicit opt in.

One thing that I'm not sure about in general is whether or not to always give the ts<->estree maps, since (as it stands) we always create them.
So this could give three levels:
- parse gives ast
- parseAndGenerate w/o project flag gives ast and maps
- parseAndGenerate w/ project flag gives ast, maps, and program

Reply:
I think for this initial PR we could start with just two: all or nothing.

- parse() will be as it is today, intended for consumers like prettier where we want an ESTree-compatible AST as quickly as possible (i.e. skip any work that isn't conversion related)

- parseAndGenerateServices() which exposes everything (ast, maps, and program), intended for consumers like ESLint who want to consume semantic/type information for advanced rule creation.

By doing it this way all the new stuff is additive and we don't have to worry about changing the experience for existing consumers. If the new stuff is not quite right for the advanced use-case, we can then further refine it in follow up PRs.

Reply:
Makes sense to me. I've exposed a separate top-level function called `parseAndGenerateServices` and made all of the additions (i.e. maps and program) only run if that new function is called _and_ the `project` option is given.